src/java/org/apache/cassandra/net/OutboundTcpConnection.java

```diff
@@ -255,6 +259,24 @@ else if (socket != null || connect())
         }
     }
 
+    @VisibleForTesting
+    int clearQueueWithFailureCallback(int currentMessage, List<QueuedMessage> bufferedMessages, int bufferSize, BlockingQueue<QueuedMessage> queue) {
+        bufferedMessages.stream().skip(currentMessage).forEach(this::invokeFailureCallback);
+        int initialCleared = bufferedMessages.size() - currentMessage;
+        bufferedMessages.clear();
+
+        int queueSize = queue.size();
+        int remaining = queueSize;
+        while (remaining > 0) {
+            remaining -= queue.drainTo(bufferedMessages, Math.min(bufferSize, remaining));
```

Comment:
Why use the same buffer as the `drainedMessageSize`? To avoid changing behavior in how many failure callbacks we could theoretically trigger at the same time?

Reply:
I figured it's nicer to keep the buffering consistent + reuse the same list, whether we're trying to actually send message or fail them we're still processing them.